module/VuFind/src/VuFind/Connection/OauthServiceTrait.php

```diff
@@ -122,14 +124,53 @@ public function authenticateWithClientCredentials(
                     return $tokenData;
                 }
             } else {
-                $this->error(
+                $this->errorSafe(
                     "Error: Nothing returned from API call."
                 );
-                $this->debug(
+                $this->debugSafe(
                     "Body return from API Call: " . print_r($body, true)
                 );
             }
         }
         return $tokenData;
     }
+
+    /**
+     * Log a debug message, if $this->log exists.
+     *
+     * @param string $msg Log message
+     *
+     * @return void
+     */
+    protected function debugSafe($msg)
```

Comment:
The problem with using the name debugSafe (as opposed to something that incorporates the name of this trait) is that if we use this technique in multiple traits, then they'll start to conflict with one another.

I'd suggest one of two possibilities:

1.) Rename these to incorporate the name of the trait (ugly, but collision-free)

2.) Instead of adding protected methods, since we only use these in one place, we could create anonymous functions inside authenticateWithClientCredentials -- e.g.:

```
$logSafe = function($level, $msg) {
    if (method_exists($this, 'log')) {
        $this->log(...func_get_args());
    }
};

$debugSafe = function ($msg) {
    $logSafe('err', $msg);
};

// ...etc
```

...and then you just change $this->debug to $debugSafe...

Reply:
Ok thanks for explaning the collision, you're right of course this is just another problem with this trait not just being able to use the LoggingAwareTrait directly.  I like your anonymous function approach but I don't want to define the methods with within authenticateWithClientCredentials, since eventually some class will probably need to use one of the other three OAuth grant types and they we'd have to define those functions yet again.  So I just renamed them to avoid the collision.